src/git-source-provider.ts

```diff
@@ -109,8 +110,17 @@ export async function getSource(settings: IGitSourceSettings): Promise<void> {
 
     // Fetch
     core.startGroup('Fetching the repository')
-    const refSpec = refHelper.getRefSpec(settings.ref, settings.commit)
-    await git.fetch(settings.fetchDepth, refSpec)
+    if (settings.fetchDepth <= 0) {
+      let refSpec = refHelper.getRefSpecForAllHistory(settings.ref)
+      await git.fetch(refSpec)
+      if (settings.commit && !(await git.shaExists(settings.commit))) {
```

Comment:
tested the shaExists fallback E2E using a force pushed branch, and using a force pushed PR

Reply:
todo: test scenario where a new commit was pushed (not force push). Going to need some additional changes to this PR to checkout the correct commit

Reply:
ok i think instead of `shaExists`, i need to test whether the remote ref matches the expected SHA. If not then override using fetch. That should fix all scenarios (force push and non force push)

Reply:
might still need `shaExists` if checking out a specific commit (no ref)